docs/testing/unit-tests.md

```diff
@@ -1,3 +1,247 @@
+                expect( mockDispatchCallback.mock.calls[ 0 ][ 0 ] ).toEqual( {
+                    type: SOME_DATA_SUCCESS,
+                    data: { hoo: 'ray!' }
+                } );
+            } );
+		} );
+     } );
+} );
+```
+
+### Mocking dependencies
+
+#### Dependency injection
+
+Passing dependencies to a function as arguments can often make your code simpler to test. As a basic example, we have a utility function that checks whether a value exists in a list:
+
+One way to do it would be to reference a dependency in a higher scope:
+
+```javascript
+import VALID_VALUES_LIST from './constants'
+
+function isValueValid( value ) {
+	return VALID_VALUES_LIST.includes( value );
+}
+
+```
+
+Our happy path test would therefore have to import and use a value from `VALID_VALUES_LIST` in order to pass:
+
+`expect( isValueValid( VALID_VALUES_LIST[ 0 ] ) ).toBe( true );`
+
+
+The above assertion is testing two behaviours: 1) that the function can detect an item in a list, and 2) that it can detect an item in `VALID_VALUES_LIST`.
+
+But what if we don't care what's stored in `VALID_VALUES_LIST`, or if the list is fetched via an HTTP request, and we only want to test whether `isValueValid` can detect an item in a list?
+
+Let's refactor the original function so that it accepts the target list as an argument:
+
+```javascript
+function isValueValid( value, validValuesList = [] ) {
+	return validValuesList.includes( value );
+}
+```
+
+Now we can pass mock lists, and, as a bonus, test a few more scenarios:
+
+`expect( isValueValid( 'hulk', [ 'batman', 'superman' ] ) ).toBe( false );`
+
+`expect( isValueValid( 'hulk', null ) ).toBe( false );`
+
+`expect( isValueValid( 'hulk', [] ) ).toBe( false );`
+
+`expect( isValueValid( 'hulk', [ 'iron man', 'hulk' ] ) ).toBe( true );`
+
+
+#### Imported dependencies
+
+Often our code will use methods and properties from imported external and internal libraries in multiple places, which makes passing around arguments messy and impracticable. For these cases `jest.mock` offers a neat way to stub these dependencies. 
+
+For instance, in Calypso, we use the ['config'](https://github.com/Automattic/wp-calypso/tree/master/client/config) module to control a great deal of functionality via feature flags. 
+
+```javascript
+// bilbo.js
+if ( config.isEnabled( 'the-ring' ) ) {
+    isBilboVisible = false;
+} else {
+	isBilboVisible = true;
+}
+```
+
+To test the behaviour under each condition, we stub the config object and use a jest mocking function to control the return value of `isEnabled`.
+
+```javascript
+// test/bilbo.js
+import { isEnabled } from 'config';
+
+jest.mock( 'config', () => {
+	const config = () => 'development';
+    
+	// default value is true
+	config.isEnabled = jest.fn( true );
+
+	return config;
+} );
+
+
+test( 'bilbo should be invisible when the `the-ring` config feature flag is disabled', () => {
+	isEnabled.mockReturnValue( false );
```

Comment:
What do you think about making this test even more `the-ring` dependant with the following code:

```js
// default value is false
 +	config.isEnabled = jest.fn( false );
```
 
In the test:

```js
expect( isBilboVisible ).toBe( true );

isEnabled.mockImplementationOnce( name => name === 'the-ring' );

expect( isBilboVisible ).toBe( false );
```

By the way, the example is not valid at the moment :)

Reply:
Just thinking out loud, rather than mocking the whole library at the top level, it might be nice to `import config from 'config';` and use `spyOn` like in the example below.

Top level mocks are both necessary and a pain in my experience.

Reply:
They are a pain, true. However, you can `spyOn` only on objects, so it isn't always a viable option.

Reply:
> What do you think about making this test even more the-ring dependant with the following code:

I see what you did there! And I'm glad you mentioned it, as it prompted me to actually create the test and run it, whereupon I found a bug. Ha!

> it might be nice to import config from 'config'; and use spyOn like in the example below

I can add an example of using `spyOn` with an object if you think it might help some contributors. Rather than do that, I might add a link to the docs somewhere if they want to read up on more on spies.